### ISSUE TITLE
Structure the model and dataloader attributes

### ISSUE THREAD

**Issue by Avsecz** [kipoi/kipoi#33]
API style:

## Model

```python
model = kipoi.Model("rbp", source="kipoi")
model.info.author
model.info.name
model.info.version
model.info.tags # functionality tags (VarEffect model)
model.info.description

model.schema.inputs
model.schema.targets
model.default_dataloader # loads the DataLoader_factory
model.predict_on_batch(x)
```

## DataLoader
```python
DataLoader = kipoi.DataLoader_factory("rbp", source="kipoi")
DataLoader.info.author
DataLoader.info.name
DataLoader.info.version
DataLoader.info.tags # functionality tags (VarEffect model)
DataLoader.info.description

DataLoader.args # argument description (dictionary)
DataLoader.schema.inputs
DataLoader.schema.metadata

dl = DataLoader(arg1="hg38.fa", intervals="file.bed")

dl.info. ...
dl.schema.inputs
dl.schema.metadata

dl[0] # first element
len(dl) # returns none on a generator
dl.batch_iter(batch_size=32, ...) # args from https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L250
# how do you know what batch-size means for the generator type
dl.load_all() # load the whole dataset
#  - if the 
```

## TODO

- [x] Define the Info class
  - Attributes:
    - author
    - name
    - version
    - tags
    - description
  - Methods:
    - `__repr__()`
    - `__str__()` - required?
- [x] Define the Schema class
  - [x] Define a list of schema classes. Each specific attributes (shape inferred from those)
    - DNASeq
    - DNASeqFwd
    - DNASeqRC (related to https://github.com/kipoi/kipoi/issues/31)
    - Array (default)
    - BigWig
    - GenomicRanges - for the metadata
  - [x] functionality to assess the compatibility between dataloaders and models
  - Common methods
    - `shape()` - shape without the batch axis
    - ?? `dtype`
- [/] `DataLoader.args` - define possible types - **Tracked in https://github.com/kipoi/kipoi/issues/45**
  - in https://github.com/kipoi/kipoi/issues/9 
    - > Programmatic way to query which inputs are needed for the pre-processor
  - mentioned in https://github.com/kipoi/kipoi/issues/31
  - types
    - SeqIntervals_file - bed file for the sequence extractor, will solve https://github.com/kipoi/kipoi/issues/11
       - [ ] do we have a better name?
    - genome_file
    - file - general file path
    - str, dict, list, int, float, ... (everything python-serializable)
   
- [x] finish "Clarify, refactor and generalize the extractor (now dataloader) specification" #21
- [x] Update the `.yaml` schema to match the class structure 1-to-1

## Questions

- [/] is `model.info.tags`, `dl.info.tags` the right word?
  - **Answer** - these will be solved by the postprocessing functionality list https://github.com/kipoi/kipoi/issues/37
  - those are functionality tags (say VarEffect model), stating that the model/dataloader supports a certain functionality and not a tag like in docker (representing the version)
    - some functionalities are inherent to model.type
       - say type(model) == "KerasModel" implements the `input_grad` and hence supports gradient-based interpretation methods
- [/] `model.info.ontology`, `model.info.classifiers`, `model.info.keywords` for ontology measures?
  - tracked in https://github.com/kipoi/kipoi/issues/23
- [/] Can we connect the `dataloader.schema` with `dataloader.args`. This is namely the input for the dataloader:
  - are there any good practices to define `dataloader.args`
  - `args -> dataloader -> output  => input -> model -> output`
  - **Answer** - unnecessary
